### PR TITLE
Implementation of the networking layer

### PR DESCRIPTION
**Pull request by gunhansancar**
This PR includes the remaining networking layer implementation in Swift

### CODE REVIEW

File: Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift [hunk 1]
```diff
@@ -6,5 +6,9 @@
 //
 
 protocol AdDataSourceType {
-    func getAd(placementId: Int, request: AdRequest, completion: @escaping Completion<Ad>)
+    func getAd(environment: Environment, placementId: Int, query: AdQuery, completion: @escaping Completion<Ad>)
```
Comment (devgabrielcoman) [inline Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift:9]: Hmm, shouldn't the `placement Id` be part of the `AdQuery`, thinking about it?

Or is:
* placement ID - part of the path
* ad query - the actual query part of the path
?

File: Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift [hunk 2]
```diff
@@ -6,5 +6,9 @@
...
+    func impression(environment: Environment, query: EventQuery, completion: @escaping Completion<Void>)
+    func click(environment: Environment, query: EventQuery, completion: @escaping Completion<Void>)
+    func videoClick(environment: Environment, query: EventQuery, completion: @escaping Completion<Void>)
+    func event(environment: Environment, query: EventQuery, completion: @escaping Completion<Void>)
```
Comment (devgabrielcoman) [inline Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift:13]: Is it worth passing the environment into every one of these methods? 
Or is it better to have it as an invariant of the `data source` ?
e.g.

```
protocol AdDataSourceType {
  var environment: Environment { get set }
  
  ....
}
```

cause theoretically you wouldn't be getting an ad on the production environment and sending its events on the staging one, no?

File: Pod/Plugin/Core/Classes/Model/EventModels.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    let ct: ConnectionType
+    let sdkVersion: String
+    let rnd: Int
+    let type: EventType?
+    let no_image: Bool?
+    let data: String?
+}
+
+struct EventRequest: Codable {
+    let environment: Environment
```
Comment (devgabrielcoman) [inline Pod/Plugin/Core/Classes/Model/EventModels.swift:22]: Following my previous comment - environment (as in the domain we're hitting) can be moved out of the event & ad request.

File: Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift (hunk 1 above)
Comment (gunhansancar) [reply Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift:9]: For the ad request, the `placementId` is the path parameter so it is not part of the query parameters.

File: Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift (hunk 2 above)
Comment (gunhansancar) [reply Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift:13]: The problem here is we let individual components e.g. `SABannerAd`, `SAInterstitialAd`, `SAVideoAd` configure the environment separately. And we have 1 datasource. We shouldn't provide the same datasource multiple times per Ad type. So moving `environment` as a field parameter wouldn't work. 

What I would suggest is having the environment set in single place as an sdk configuration param on the init of the sdk.  However then, we would lose to ability to individually set environment per ad type but I don't think we actually need it or anyone else needs that capability.

Comment (devgabrielcoman) [reply Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift:13]: ^^ yes, we don't need it on a per ad type. in a given session, the environment should be the same for all ads. And it should be a config param of sorts 👍

File: Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift (hunk 1 above)
Comment (devgabrielcoman) [reply Pod/Plugin/Core/Classes/DataSources/AdDataSource.swift:9]: agreed 👍

**Pull request merged**